### ISSUE TITLE
wamp.session.on_leave doesn't provide session ID

### ISSUE THREAD

**Issue by Budulianin** [crossbario/crossbar#451]
crossbar==0.11.0

wamp.session.on_join pass info about session, but on_leave not.

`def on_join(self, session):
    print(session)`

`def on_leave(self, *args, **kwargs):
    print(args, kwargs)`

output:
`{'authprovider': None, ... 'type': 'websocket'}}`
`((None,), {})`

config crossbar:

```
{
  "controller": {
  },
  "workers": [
    {
      "type": "router",
      "options": {
        "pythonpath": [".."]
      },
      "realms": [
        {
          "name": "general",
          "roles": [
            {
               "name": "anonymous",
               "permissions": [
                  {
                     "uri": "*",
                     "publish": true,
                     "subscribe": true,
                     "call": true,
                     "register": true
                  }
               ]
            },
            {
              "name": "backend",
              "permissions": [
                {
                  "uri": "*",
                  "publish": true,
                  "subscribe": true
                }
              ]
            }
          ]
        }
      ],
      "transports": [
        {
          "type": "websocket",
          "endpoint": {
            "type": "tcp",
            "port": 6060
          },
          "options": {
            "enable_webstatus": false
          }
        }
      ],
      "components": [
        {
          "type": "class",
          "classname": "directory.Application",
          "realm": "general",
          "role": "backend"
        }
      ]
    }
  ]
}
```

**Comment by oberstet** [comment]
correct. pls see: https://github.com/tavendo/WAMP/blob/master/spec/advanced/session-meta-api.md#wampsessionon_leave

**Comment by Budulianin** [comment]
wamp.session.on_leave:
`Event Arguments
session | id`

When fire wamp.session.on_join, then sessionId exists, but when fire wamp.session.on_leave, sessionId is None. Why sessionId is None if it existed?

**Comment by oberstet** [comment]
oh, if this is the case, that is a bug. reopening. I have to look.

**Comment by Budulianin** [comment]
>oh, if this is the case, that is a bug. reopening. I have to look.
Yes it's case. I need to register some information from session when user came and delete this information when user left. At the moment I can use for this `wamp.session.on_join` but wamp.session.on_leave can't, it's strange and not logical. By the title I thought that `wamp.session.on_leave` has the same behavior as `wamp.session.on_join` and it's logic in my opinion.

**Comment by oberstet** [comment]
`wamp.session.on_leave` should provide the session ID, but only that.

**Comment by Budulianin** [comment]
>should provide the session ID

But at the moment, `wamp.session.on_leave` pass in handler None.

**Comment by Webbaroli** [comment]
Hi guys! I experience the same problem. I think the solution is at line 390 of router/session.py
you should fire the callback BEFORE setting session id to null. Is it the correct way?

                self._session_id = None
                self._pending_session_id = None

                # fire callback and close the transport
                self.onLeave(types.CloseDetails(msg.reason, msg.message))

**Comment by adamlc** [comment]
I too have just come across this issue :cry:

**Comment by stopthatcow** [comment]
As have I. :(

**Comment by stopthatcow** [comment]
It looks like this only affects web-socket based sessions. I get the proper behavior using a raw TCP session.

**Comment by meejah** [comment]
It's not quite as simple as moving the call I don't believe; there are comments in there (and a test-case) ensuring that _session_id is None before onLeave gets called in that flow (relating to #431 it seems).

**Comment by adamlc** [comment]
I also get the same behaviour as @stopthatcow, it works for raw sessions but not websockets.

**Comment by meejah** [comment]
If anyone experiencing this problem would like to test their code with the fix proposed in #520, that would be great!

You can install a crossbar with that fix like so: ``pip install https://github.com/meejah/crossbar/archive/issue451.zip`` (it is based off current master).
